core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1434,8 +1434,15 @@ class KafkaApis(val requestChannel: RequestChannel,
       return
     }
 
-    def sendResponseCallback(pid: Long)(responseStatus: Map[TopicPartition, PartitionResponse]): Unit = {
+    def sendResponseCallback(pid: Long, result: TransactionResult)(responseStatus: Map[TopicPartition, PartitionResponse]): Unit = {
       errors.put(pid, responseStatus.mapValues(_.error).asJava)
+
+      val successfulPartitions = responseStatus.filter { case (_, partitionResponse) =>
+        partitionResponse.error == Errors.NONE
+      }.keys.toSeq
+
+      groupCoordinator.handleTxnCompletion(producerId = pid, topicPartitions = successfulPartitions, transactionResult = result)
```

Comment:
Discussed offline. At the moment, the call to `handleTxnCompletion` seems safe, but to be resilient to future changes, we should probably add code to catch any exceptions and overwrite the response error codes with an UNKNOWN error. This should cause the TC to retry.